components/common/src/main/java/com/hotels/styx/common/content/QueueDrainingPublisher.java

```diff
@@ -0,0 +1,55 @@
+/*
+  Copyright (C) 2013-2020 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.common.content;
+
+import com.hotels.styx.api.Buffer;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.concurrent.Executor;
+
+/**
+ * A publisher to wrap the FlowControllingHttpContentProducer FSM and perform subscription operations via a QueueDrainingExecutor.
+ */
+public final class QueueDrainingPublisher implements Publisher<Buffer> {
```

Comment:
This class works with any executor. Apart from name,  there is nothing specific to queue draining event processor. Therefore you could call `queueDrainingExecutor` just `delegate`, etc. and perhaps rename this class to something more generic like `FlowControllingSubscriptionAdapter` etc? Or something better, actually ;-)